src/urlencoded.rs

```diff
@@ -1,11 +1,137 @@
 use groupable::Groupable;
+use hyper::uri::RequestUri;
+use hyper::uri::RequestUri::{Star, AbsoluteUri, AbsolutePath, Authority};
+use plugin::{Plugin, Pluggable};
+use request::Request;
+use status::StatusCode;
+use std::collections::HashMap;
+use std::io::{Error, Read};
+use typemap::Key;
+use url::{form_urlencoded, UrlParser};
+
+type QueryStore = HashMap<String, Vec<String>>;
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Query(QueryStore);
+
+impl Query {
+    /// Retrieves the first value from the query for `key`, or `None` if not present.
+    ///
+    /// # Notes
+    /// There may be multiple values per key, if all of the values for a given
+    /// `key` are required, then use `all`.
+    //FIXME: Implement via Indexing whenever IndexGet is supported
+    pub fn get(&self, key: &str) -> Option<&str> {
+        self.0.get(key).and_then(|v| v.first().map(|s| &**s))
+    }
 
-pub fn parse (encoded_string : &str) -> HashMap<String, Vec<String>> {
+    /// Retrieve all values from the query for `key`, or `None` if none are present.
+    pub fn all(&self, key: &str) -> Option<&[String]> {
+        self.0.get(key).map(|v| &**v)
+    }
+}
+
+fn parse (encoded_string : &str) -> QueryStore {
     form_urlencoded::parse(encoded_string.as_bytes()).into_iter().group()
 }
 
+fn parse_uri(origin: &RequestUri) -> Query {
+    let f = |query: Option<&String>| query.map(|q| parse(&*q));
+
+    let result = match *origin {
+        AbsoluteUri(ref url) => f(url.query.as_ref()),
+        AbsolutePath(ref s) => UrlParser::new().parse_path(&*s)
+                                                // FIXME: If this fails to parse,
+                                                // then it really shouldn't have
+                                                // reached here.
+                                               .ok()
+                                               .and_then(|(_, query, _)| f(query.as_ref())),
+        Star | Authority(..) => None
+    };
+
+    Query(result.unwrap_or_else(|| HashMap::new()))
+}
+
+struct FormBodyParser;
+
+impl Key for FormBodyParser {
+    type Value = Query;
+}
+
+impl<'mw, 'conn, D> Plugin<Request<'mw, 'conn, D>> for FormBodyParser {
+    type Error = Error;
+
+    fn eval(req: &mut Request<D>) -> Result<Query, Error> {
+        let mut buf = String::new();
+        try!(req.origin.read_to_string(&mut buf));
+        Ok(Query(parse(&*buf)))
+    }
+}
+
+pub trait FormBody {
+    /// Extracts URL encoded data from the request body.
+    /// # Examples
+    /// ```{rust}
+    /// #[macro_use] extern crate nickel;
+    /// use nickel::{Nickel, HttpRouter, FormBody};
+    ///
+    /// fn main() {
+    ///     let mut server = Nickel::new();
+    ///     server.post("/a", middleware! { |req, res|
+    ///         let post = try_with!(res, req.post());
+    ///         return res.send(format!("Post: {:?}", post))
+    ///     });
+    /// }
+    /// ```
+    fn post(&mut self) -> Result<&Query, (StatusCode, Error)>;
```

Comment:
Yeah, I just don't think using `body` is a good idea yet until we can handle more things, hence `form_body` :P Sorry for causing confusion, thanks for taking care of it :+1: